Serialize schema to a file

mansu opened the pull request:
Move `FieldDef` to `LuceneFieldDef` into metadata.proto. Add a new metadata proto file for storing schema. Add serialization code and related unit tests. Also refactored `SchemaAwareLogDocumentBuilderImpl.java. Update `LogBuilder` to use the schema. 

This code is part of #389. Refactored it so it's easier to review.

vthacker commented on kaldb/src/main/java/com/slack/kaldb/logstore/DocumentBuilder.java:
```diff
@@ -1,8 +1,12 @@
 public interface DocumentBuilder<T> {
   Document fromMessage(T message) throws IOException;
+
+  Map<String, LuceneFieldDef> getSchema();
```
does this need to be public?

mansu replied:
Interfaces are public by default.

vthacker commented on kaldb/src/main/java/com/slack/kaldb/logstore/schema/SchemaAwareLogDocumentBuilderImpl.java:
```diff
@@ -0,0 +1,344 @@
+ * rarely an issue and helps with performance. If this is an issue, we need to scan the json twice
+ * to ensure document is good to index.
+ */
+public class SchemaAwareLogDocumentBuilderImpl implements DocumentBuilder<LogMessage> {
```
can we not change the package name in this PR? Or let's do it in a previous PR?

It's impossible to tell what changed in this class since the package was moved and also the number of lines here is much lesser.

So to make it a little more friendly to review can we wither list down the changes from the previous impl, or do the package change in a separate PR?

mansu replied:
Done.

vthacker commented:
LuceneFieldDef proto has storeNumericDocValue, but it should just be storeDocValues (not numeric specific). We can add doc-values to string fields and that's the only way to do basic group by on a string field like hostname. Today we can't do that with the current design.

I'd like to see LuceneFieldDef take the field type enum instead of a string. That's looks safer to me. If not we should add validation in the ctor of LuceneFieldDef and some tests to validate it.

I think we should split TextField into two - How it's currently implemented is confusing and we can make it simpler by splitting it into two - TextField and StringField
- TextField is a field that can be indexed/stored/analyzed and never have doc values
- StringField can be indexed and have doc values but never be analyzed.

I like the idea of adding a metadata map to ChunkSchema however I'm curious to learn if you have an use-case in mind for it? Or is it just a good to have sort of thing?

Can we add some unit tests for SchemaAwareLogDocumentBuilderImpl#getJsonType that tests how it chooses LONG vs INT correctly?

Is the name SchemaAwareLogDocumentBuilderImpl is confusing? It's not per document right? 

Here is a test case that potentially exposes a bug? If I am understanding correctly Since the test is building both document using SchemaAwareLogDocumentBuilderImpl, the code should be able track that "intField" was indexed with a numeric value for document 1 and string value for document 2?

```
@Test
  public void testSameFieldDifferentValueTypes() throws JsonProcessingException {
    SchemaAwareLogDocumentBuilderImpl docBuilder = build(CONVERT_FIELD_VALUE, meterRegistry);
    assertThat(docBuilder.getIndexFieldConflictPolicy()).isEqualTo(CONVERT_FIELD_VALUE);
    assertThat(docBuilder.getSchema().size()).isEqualTo(17);

    LogMessage msg1 =
            new LogMessage(
                    MessageUtil.TEST_DATASET_NAME,
                    "INFO",
                    "1",
                    Map.of(
                            LogMessage.ReservedField.TIMESTAMP.fieldName,
                            MessageUtil.getCurrentLogDate(),
                            LogMessage.ReservedField.MESSAGE.fieldName,
                            "Test message1",
                    "intField", 1));
    Document testDocument1 = docBuilder.fromMessage(msg1);
    assertThat(testDocument1.getFields().size()).isEqualTo(10);

    LogMessage msg2 =
            new LogMessage(
                    MessageUtil.TEST_DATASET_NAME,
                    "INFO",
                    "2",
                    Map.of(
                            LogMessage.ReservedField.TIMESTAMP.fieldName,
                            MessageUtil.getCurrentLogDate(),
                            LogMessage.ReservedField.MESSAGE.fieldName,
                            "Test message2",
                            "intField", "not_an_int"));
    Document testDocument2 = docBuilder.fromMessage(msg2);
    // Potential bug? This shoudn't be 10 right? now the same intField should be converted to a text field? We use CONVERT_FIELD_VALUE in the test
    // Also if you look closely to testDocument3 it shows the intField is being indexed with value 0 as a pointField and DocValue Field
    assertThat(testDocument2.getFields().size()).isEqualTo(11);
  }
  ```

bryanlb commented on kaldb/src/main/java/com/slack/kaldb/metadata/schema/ChunkSchema.java:
```diff
@@ -0,0 +1,65 @@
+import java.nio.file.Files;
+import java.util.Map;
+
+/** Schema for a chunk. */
```
This needs better class documentation around what the purpose of the metadata is for, who uses it, and why this class is needed.

bryanlb commented on kaldb/src/main/java/com/slack/kaldb/metadata/schema/FieldType.java:
```diff
@@ -0,0 +1,228 @@
+import org.apache.lucene.document.StringField;
+import org.apache.lucene.document.TextField;
+
+public enum FieldType {
```
This enum should have documentation.

bryanlb commented on kaldb/src/main/java/com/slack/kaldb/metadata/schema/LuceneFieldDef.java:
```diff
@@ -0,0 +1,82 @@
+    super(name);
+    this.fieldType = FieldType.valueOf(fieldType.trim().toUpperCase());
+    if (isAnalyzed && !isIndexed) {
+      throw new InvalidFieldDefException("Cannot set isAnalyzed without setting isIndexed");
```
This should be using `Preconditions` checks, (ie, `checkArgument`), as that is the pattern used for all of the other `KaldbMetadata` classes.

mansu commented:
> LuceneFieldDef proto has storeNumericDocValue, but it should just be storeDocValues (not numeric specific). We can add doc-values to string fields and that's the only way to do basic group by on a string field like hostname. Today we can't do that with the current design.
Updated the variable to StoreDocValues.

> 
> I'd like to see LuceneFieldDef take the field type enum instead of a string. That's looks safer to me. If not we should add validation in the ctor of LuceneFieldDef and some tests to validate it.
>
This code would also power the config eventually. So, kept it as a string field. Added tests to make sure a invalid type thows exception. 
> I think we should split TextField into two - How it's currently implemented is confusing and we can make it simpler by splitting it into two - TextField and StringField

> TextField is a field that can be indexed/stored/analyzed and never have doc values
> StringField can be indexed and have doc values but never be analyzed.
Added new String field type. Kept text as a separate field.

> I like the idea of adding a metadata map to ChunkSchema however I'm curious to learn if you have an use-case in mind for it? Or is it just a good to have sort of thing?
For now, it will help us add more schema metadata without making code changes. 

> Can we add some unit tests for SchemaAwareLogDocumentBuilderImpl#getJsonType that tests how it chooses LONG vs INT correctly?
> 
In JSON, we can't have int's everything is a Long. 

> Is the name SchemaAwareLogDocumentBuilderImpl is confusing? It's not per document right?
> 
The fromMessage method builds a document out of a message. The name is inline with our other classes. 

> Here is a test case that potentially exposes a bug? If I am understanding correctly Since the test is building both document using SchemaAwareLogDocumentBuilderImpl, the code should be able track that "intField" was indexed with a numeric value for document 1 and string value for document 2?
> 
> @Test
>   public void testSameFieldDifferentValueTypes() throws JsonProcessingException {
>     SchemaAwareLogDocumentBuilderImpl docBuilder = build(CONVERT_FIELD_VALUE, meterRegistry);
>     assertThat(docBuilder.getIndexFieldConflictPolicy()).isEqualTo(CONVERT_FIELD_VALUE);
>     assertThat(docBuilder.getSchema().size()).isEqualTo(17);
> 
>     LogMessage msg1 =
>             new LogMessage(
>                     MessageUtil.TEST_DATASET_NAME,
>                     "INFO",
>                     "1",
>                     Map.of(
>                             LogMessage.ReservedField.TIMESTAMP.fieldName,
>                             MessageUtil.getCurrentLogDate(),
>                             LogMessage.ReservedField.MESSAGE.fieldName,
>                             "Test message1",
>                     "intField", 1));
>     Document testDocument1 = docBuilder.fromMessage(msg1);
>     assertThat(testDocument1.getFields().size()).isEqualTo(10);
> 
>     LogMessage msg2 =
>             new LogMessage(
>                     MessageUtil.TEST_DATASET_NAME,
>                     "INFO",
>                     "2",
>                     Map.of(
>                             LogMessage.ReservedField.TIMESTAMP.fieldName,
>                             MessageUtil.getCurrentLogDate(),
>                             LogMessage.ReservedField.MESSAGE.fieldName,
>                             "Test message2",
>                             "intField", "not_an_int"));
>     Document testDocument2 = docBuilder.fromMessage(msg2);
>     // Potential bug? This shoudn't be 10 right? now the same intField should be converted to a text field? We use CONVERT_FIELD_VALUE in the test
>     // Also if you look closely to testDocument3 it shows the intField is being indexed with value 0 as a pointField and DocValue Field
>     assertThat(testDocument2.getFields().size()).isEqualTo(11);
>   }
There is no bug in this code. Were you thinking about CONVERT_AND_DUPLICATE_FIELD? That is different than CONVERT_FIELD_VALUE which only converts the field value.

vthacker commented:
```
@Test
  public void testSameFieldDifferentValueTypes() throws JsonProcessingException {
    SchemaAwareLogDocumentBuilderImpl docBuilder = build(CONVERT_FIELD_VALUE, meterRegistry);
    assertThat(docBuilder.getIndexFieldConflictPolicy()).isEqualTo(CONVERT_FIELD_VALUE);
    assertThat(docBuilder.getSchema().size()).isEqualTo(17);

    LogMessage msg1 =
            new LogMessage(
                    MessageUtil.TEST_DATASET_NAME,
                    "INFO",
                    "1",
                    Map.of(
                            LogMessage.ReservedField.TIMESTAMP.fieldName,
                            MessageUtil.getCurrentLogDate(),
                            LogMessage.ReservedField.MESSAGE.fieldName,
                            "Test message1",
                            "intField", 1));
    Document testDocument1 = docBuilder.fromMessage(msg1);
    assertThat(testDocument1.getFields().size()).isEqualTo(10);

    LogMessage msg2 =
            new LogMessage(
                    MessageUtil.TEST_DATASET_NAME,
                    "INFO",
                    "2",
                    Map.of(
                            LogMessage.ReservedField.TIMESTAMP.fieldName,
                            MessageUtil.getCurrentLogDate(),
                            LogMessage.ReservedField.MESSAGE.fieldName,
                            "Test message2",
                            "intField", "not_an_int"));
    Document testDocument2 = docBuilder.fromMessage(msg2);
    // Potential bug? This shoudn't be 10 right? now the same intField should be converted to a text field? We use CONVERT_FIELD_VALUE in the test
    // Also if you look closely to testDocument3 it shows the intField is being indexed with value 0 as a pointField and DocValue Field
    assertThat(testDocument2.getFields().size()).isEqualTo(10);
  }
```

This test still fails - We should be getting 10 according to our chat but it's 12

bryanlb commented on kaldb/src/test/java/com/slack/kaldb/logstore/schema/SchemaAwareLogDocumentBuilderImplTest.java:
```diff
@@ -7,16 +7,17 @@
-import static com.slack.kaldb.logstore.schema.SchemaAwareLogDocumentBuilderImpl.FieldType.convertFieldValue;
 import static com.slack.kaldb.logstore.schema.SchemaAwareLogDocumentBuilderImpl.build;
 import static com.slack.kaldb.logstore.schema.SchemaAwareLogDocumentBuilderImpl.makeNewFieldOfType;
+import static com.slack.kaldb.metadata.schema.FieldType.convertFieldValue;
```
Now that this code has been moved the corresponding tests should be moved in the appropriate class as well. We're testing `FieldType` static methods in the `SchemaAwareLogDocumentBuilderImplTest.java` class, when the `SchemaAwareLogDocumentBuilderImpl` no longer contains this code.

mansu replied:
Moved.

bryanlb commented on kaldb/src/main/java/com/slack/kaldb/logstore/schema/SchemaAwareLogDocumentBuilderImpl.java:
```diff
@@ -48,285 +42,55 @@ public class SchemaAwareLogDocumentBuilderImpl implements DocumentBuilder<LogMes
-
-      return null;
-    }
+  private static void addTextField(
```
Doesn't look like this was removed as mentioned.

bryanlb commented on kaldb/src/main/java/com/slack/kaldb/metadata/schema/ChunkSchema.java:
```diff
@@ -0,0 +1,74 @@
+ * Since this field is written and read my multiple threads it is hardcoded as a ConcurrentHashMap.
+ * New fields are added to the fieldMap when it's written and read during query.
+ */
+public class ChunkSchema extends KaldbMetadata {
```
We seem to be missing the corresponding `ChunkSchemaStore` for persisting this metadata, and the associated test.

bryanlb commented on kaldb/src/main/java/com/slack/kaldb/metadata/schema/LuceneFieldDef.java:
```diff
@@ -0,0 +1,57 @@
+/*
+ * LuceneFieldDef describes the configs that can be set on a lucene field. This config defines how a field is indexed.
+ */
+public class LuceneFieldDef extends KaldbMetadata {
```
We seem to be missing the corresponding metadata stores and associated tests.

bryanlb commented on kaldb/src/main/java/com/slack/kaldb/logstore/LogDocumentBuilderImpl.java:
```diff
@@ -370,4 +372,9 @@ public Document fromMessage(LogMessage message) throws JsonProcessingException {
+
+  @Override
+  public Map<String, LuceneFieldDef> getSchema() {
+    return Collections.emptyMap();
```
This should have a comment explaining why an empty map is appropriate here.

mansu replied:
Fixed.

bryanlb commented on kaldb/src/main/java/com/slack/kaldb/logstore/schema/SchemaAwareLogDocumentBuilderImpl.java:
```diff
@@ -348,30 +135,29 @@ public enum FieldConflictPolicy {
   }
 
-  private static final Map<FieldType, FieldDef> defaultPropDescriptionForType =
+  private static final String DUMMY_FIELD = "DUMMY_FIELD";
```
This needs a comment explaining why a dummy field is being used here for the default properties.

bryanlb commented on kaldb/src/main/proto/metadata.proto:
```diff
@@ -147,3 +147,29 @@ message DatasetPartitionMetadata {
+  string type = 2;
+
+  // Store the value as a Lucene StoredField.
+  bool isStored = 3;
```
These fields aren't using the correct naming convention, which will fail linting - see https://github.com/slackhq/kaldb/pull/222

bryanlb commented on kaldb/src/main/proto/metadata.proto:
```diff
@@ -147,3 +147,29 @@ message DatasetPartitionMetadata {
+  // Name of the chunk.
+  string name = 1;
+  // Field map for each field in the lucene index in the chunk.
+  map<string, LuceneFieldDef> fieldDefMap = 2;
```
This probably shouldn't be suffixed map, since none of the other maps are suffixed map - even in the same class.

mansu replied:
Tried removing the map, but the code is less readable. This is a map from an app perspective for field definitions and not meant to reflect the type (which also happens to match). So leaving it as is.

mansu commented on kaldb/src/main/java/com/slack/kaldb/metadata/schema/ChunkSchema.java:
```diff
@@ -0,0 +1,74 @@
+ * Since this field is written and read my multiple threads it is hardcoded as a ConcurrentHashMap.
+ * New fields are added to the fieldMap when it's written and read during query.
+ */
+public class ChunkSchema extends KaldbMetadata {
```
Add tests.
